File: java/src/main/java/accounting/patterns/Customer.java
package accounting.patterns;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

public class Customer {
    private final String name;
    private final ServiceAgreement serviceAgreement;
    private final List<Entry> entries = new ArrayList<>();

    public Customer(String name, ServiceAgreement serviceAgreement) {
        this.name = name;
        this.serviceAgreement = serviceAgreement;
    }

    public String getName() {
        return name;
    }

    public Entry getEntry(int index) {
        return entries.get(index);
    }

    public void post(EventType eventType, LocalDate eventDate, Quantity quantity) {
        entries.add(serviceAgreement.post(eventType, eventDate, quantity));
    }

    public MonetaryAmount balance(LocalDate balanceDate) {
        return entries.stream()
                .filter(entry -> entry.getEntryDate().isBefore(balanceDate))
                .map(Entry::getAmount)
                .reduce(MonetaryAmount.ZERO, MonetaryAmount::add);
    }
}

File: src/main/typescript/accounting-pattern/Customer.ts
import type { Entry } from "./Entry";
import type { EventType } from "./EventType";
import { MonetaryAmount } from "./MonetaryAmount";
import type { Quantity } from "./Quantity";
import type { ServiceAgreement } from "./ServiceAgreement";

/** An account holder, its service agreement, and the entries posted to it. */
export class Customer {
    private readonly name: string;
    private readonly serviceAgreement: ServiceAgreement;
    private readonly entries: Entry[] = [];

    constructor(name: string, serviceAgreement: ServiceAgreement) {
        this.name = name;
        this.serviceAgreement = serviceAgreement;
    }

    getName(): string {
        return this.name;
    }

    getEntry(index: number): Entry {
        return this.entries[index];
    }

    post(eventType: EventType, eventDate: Date, quantity: Quantity): void {
        this.entries.push(this.serviceAgreement.post(eventType, eventDate, quantity));
    }

    balance(balanceDate: Date): MonetaryAmount {
        return this.entries
            .filter(entry => entry.getEntryDate() < balanceDate)
            .map(entry => entry.getAmount())
            .reduce((total, amount) => total.add(amount), MonetaryAmount.ZERO);
    }
}
